tfhe/src/shortint/keycache.rs

```diff
@@ -396,7 +438,78 @@ impl KeycacheWopbsV0 {
     }
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct KeySwitchingKeyParams(
+    pub PBSParameters,
+    pub PBSParameters,
+    pub ShortintKeySwitchingParameters,
+);
+
+impl<P> From<(P, P, ShortintKeySwitchingParameters)> for KeySwitchingKeyParams
+where
+    P: Into<PBSParameters>,
+{
+    fn from(tuple: (P, P, ShortintKeySwitchingParameters)) -> Self {
+        Self(tuple.0.into(), tuple.1.into(), tuple.2)
+    }
+}
+
+impl From<KeySwitchingKeyParams> for KeySwitchingKey {
+    fn from(params: KeySwitchingKeyParams) -> Self {
+        // use with_key to avoid doing a temporary cloning
+        KEY_CACHE.inner.with_key(params.0, |keys_1| {
+            KEY_CACHE.inner.with_key(params.1, |keys_2| {
+                KeySwitchingKey::new((&keys_1.0, &keys_1.1), (&keys_2.0, &keys_2.1), params.2)
+            })
+        })
+    }
+}
+
+impl NamedParam for KeySwitchingKeyParams {
+    fn name(&self) -> String {
+        format!("{}__{}__{}", self.0.name(), self.1.name(), self.2.name())
+    }
+}
```

Comment:
can this macroized ?

Reply:
using paste and the exposed constants for parameter structs ?

Reply:
This can't be macroized as some parameters (mainly ShortintKeySwitchingParameters, (self.2.name()) may be computed at runtime